File: scripts/forks/replay_ledger.py
#!/usr/bin/env python3
"""Committed replay ledger support for Forks JSON submissions."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import forks

LEDGER_FORMAT = "LS_FORK_REPLAY_LEDGER_V1"
ENTRY_FORMAT = "LS_FORK_REPLAY_LEDGER_ENTRY_V1"
LEDGER_ROOT = "forks/replay-ledger"


def safe_token(raw: str) -> str:
    value = str(raw).strip().replace("\\", "/")
    if not value or value in {".", ".."}:
        raise RuntimeError(f"unsafe ledger token: {raw!r}")
    if ":" in value:
        raise RuntimeError(f"unsafe ledger token: {raw!r}")
    parts = value.split("/")
    if any(part in {"", ".", ".."} for part in parts):
        raise RuntimeError(f"unsafe ledger token: {raw!r}")
    return "_".join(parts)


def canonical_json(value: Any) -> str:
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def json_sha256(value: Any) -> str:
    return forks.sha256_text(canonical_json(value))


def target_descriptor(payload: dict[str, Any], target_ref: str) -> dict[str, Any]:
    target = payload.get("target")
    if isinstance(target, dict):
        kind = target.get("kind", "gadget")
        if kind == "gadget" and isinstance(target.get("gizmo"), str):
            gadget = target.get("gadget") or target.get("lane")
            if isinstance(gadget, str) and gadget:
                return {
                    "kind": "gadget",
                    "gizmo": target["gizmo"],
                    "gadget": gadget,
                    "target_ref": target_ref,
                }
    return {
        "kind": "ref",
        "target_ref": target_ref,
    }


def ledger_relpath(agent: str, target: dict[str, Any]) -> str:
    clean_agent = safe_token(agent)
    if target.get("kind") == "gadget":
        gizmo = safe_token(str(target.get("gizmo", "")))
        gadget = safe_token(str(target.get("gadget", "")))
        return f"{LEDGER_ROOT}/gadgets/{gizmo}/{gadget}/{clean_agent}.json"
    ref = safe_token(str(target.get("target_ref", "unknown")))
    return f"{LEDGER_ROOT}/refs/{ref}/{clean_agent}.json"


def file_fingerprint(item: dict[str, Any]) -> dict[str, Any]:
    op = item.get("op")
    path = item.get("path")
    out: dict[str, Any] = {
        "op": op,
        "path": path,
    }
    if op == "upsert":
        content = item.get("content")
        encoding = item.get("encoding", "utf-8")
        out["encoding"] = encoding
        out["content_sha256"] = forks.sha256_text(content if isinstance(content, str) else "")
        out["content_length"] = len(content) if isinstance(content, str) else None
    return out


def patch_fingerprint(payload: dict[str, Any]) -> dict[str, Any]:
    files = payload.get("files", [])
    return {
        "json_patch_sha256": json_sha256(payload),
        "title": payload.get("title", ""),
        "agent": payload.get("agent"),
        "file_count": len(files) if isinstance(files, list) else 0,
        "files": [file_fingerprint(item) for item in files] if isinstance(files, list) else [],
    }


def empty_ledger(agent: str, target: dict[str, Any]) -> dict[str, Any]:
    return {
        "format": LEDGER_FORMAT,
        "agent": forks.normalize_agent(agent),
        "target": target,
        "created_at": forks.now_iso(),
        "updated_at": forks.now_iso(),
        "next_sequence": 1,
        "entries": [],
    }


def read_ledger(path: Path, agent: str, target: dict[str, Any]) -> dict[str, Any]:
    if not path.exists():
        return empty_ledger(agent, target)
    data = json.loads(path.read_text(encoding="utf-8"))
    if data.get("format") != LEDGER_FORMAT:
        raise RuntimeError(f"invalid replay ledger format: {path}")
    if data.get("agent") != forks.normalize_agent(agent):
        raise RuntimeError(f"replay ledger agent mismatch: {path}")
    if not isinstance(data.get("entries"), list):
        raise RuntimeError(f"invalid replay ledger entries: {path}")
    if not isinstance(data.get("next_sequence"), int):
        data["next_sequence"] = len(data["entries"]) + 1
    return data


def find_entry(ledger: dict[str, Any], json_patch_sha256: str) -> dict[str, Any] | None:
    for entry in ledger.get("entries", []):
        if entry.get("json_patch_sha256") == json_patch_sha256:
            return entry
    return None


def append_entry(work: Path, agent: str, payload: dict[str, Any], target_ref: str) -> dict[str, Any]:
    agent = forks.normalize_agent(agent)
    target = target_descriptor(payload, target_ref)
    relpath = ledger_relpath(agent, target)
    path = work / relpath
    fingerprint = patch_fingerprint(payload)
    ledger = read_ledger(path, agent, target)
    existing = find_entry(ledger, fingerprint["json_patch_sha256"])
    if existing is not None:
        return {
            "path": relpath,
            "entry": existing,
            "appended": False,
            "reason": "json patch fingerprint already present",
        }
    sequence = int(ledger.get("next_sequence", len(ledger.get("entries", [])) + 1))
    entry = {
        "format": ENTRY_FORMAT,
        "sequence": sequence,
        "agent": agent,
        "target": target,
        "title": fingerprint["title"],
        "json_patch_sha256": fingerprint["json_patch_sha256"],
        "file_count": fingerprint["file_count"],
        "file_fingerprints": fingerprint["files"],
        "target_ref_at_capture": target_ref,
        "created_at": forks.now_iso(),
    }
    ledger["target"] = target
    ledger["updated_at"] = forks.now_iso()
    ledger["next_sequence"] = sequence + 1
    ledger["entries"].append(entry)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(ledger, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    return {
        "path": relpath,
        "entry": entry,
        "appended": True,
    }


def ledger_fingerprint(ledger: dict[str, Any]) -> str:
    stable = {
        "format": ledger.get("format"),
        "agent": ledger.get("agent"),
        "target": ledger.get("target"),
        "entries": ledger.get("entries", []),
    }
    return json_sha256(stable)

File: projects/lambda-script/src/lambda-script/lib/LambdaScript/Forks/ReplayLedger.hs
{-# LANGUAGE OverloadedStrings #-}

-- | Committed replay ledger (ported from scripts/forks/replay_ledger.py).
--
-- When a JSON patch lands, its payload is stored content-addressed and an entry
-- is appended to the per-agent ledger, both written into the candidate worktree
-- so they are committed onto the integration branch. This is what makes an agent
-- lane disposable: the work is preserved (and later audited as materialised)
-- in the ledger, independent of the lane branch.
module LambdaScript.Forks.ReplayLedger
  ( appendEntry
  , gadgetLedgerRelpath
  ) where

import Data.Aeson (Value(..), encode, eitherDecodeFileStrict', object, (.=))
import qualified Data.Aeson.Key as K
import qualified Data.Aeson.KeyMap as KM
import qualified Data.ByteString.Lazy as BL
import Data.Foldable (toList)
import qualified Data.Scientific as Sci
import qualified Data.Text as T
import System.Directory (createDirectoryIfMissing, doesFileExist)
import System.FilePath ((</>), takeDirectory)

import LambdaScript.Forks.Git (normalizeAgent, nowIso)
import LambdaScript.Forks.Hash (jsonSha256, sha256Text)

ledgerRoot :: String
ledgerRoot = "forks/replay-ledger"

payloadRoot :: String
payloadRoot = ledgerRoot ++ "/payloads"

ledgerFormat, entryFormat, payloadFormat :: String
ledgerFormat = "LS_FORK_REPLAY_LEDGER_V1"
entryFormat = "LS_FORK_REPLAY_LEDGER_ENTRY_V1"
payloadFormat = "LS_FORK_REPLAY_PAYLOAD_V1"

-- helpers ---------------------------------------------------------------------

objLookup :: Value -> String -> Maybe Value
objLookup (Object o) k = KM.lookup (K.fromString k) o
objLookup _ _ = Nothing

objStr :: Value -> String -> Maybe String
objStr v k =
  case objLookup v k of
    Just (String s) -> Just (T.unpack s)
    _ -> Nothing

asInt :: Value -> Maybe Int
asInt (Number n) = Sci.toBoundedInteger n
asInt _ = Nothing

asArray :: Maybe Value -> [Value]
asArray (Just (Array a)) = toList a
asArray _ = []

-- collapse path separators / colons into a single token (mirrors safe_token join)
safeToken :: String -> String
safeToken = map (\c -> if c == '/' || c == '\\' || c == ':' then '_' else c)

payloadRelpath :: String -> String
payloadRelpath sha = payloadRoot ++ "/" ++ take 2 sha ++ "/" ++ sha ++ ".json"

-- target descriptor + ledger location ----------------------------------------

targetDescriptor :: Value -> String -> Value
targetDescriptor payload targetRef =
  case objLookup payload "target" of
    Just t@(Object _) ->
      let kind = maybe "gadget" id (objStr t "kind")
          gizmo = objStr t "gizmo"
          gadget = case objStr t "gadget" of
                     Just g -> Just g
                     Nothing -> objStr t "lane"
      in case (kind, gizmo, gadget) of
           ("gadget", Just gz, Just gd) ->
             object ["kind" .= ("gadget" :: String), "gizmo" .= gz, "gadget" .= gd, "target_ref" .= targetRef]
           _ -> refTarget
    _ -> refTarget
  where
    refTarget = object ["kind" .= ("ref" :: String), "target_ref" .= targetRef]

ledgerRelpath :: String -> Value -> String
ledgerRelpath agent target =
  case objStr target "kind" of
    Just "gadget" ->
      gadgetLedgerRelpath (maybe "" id (objStr target "gizmo")) (maybe "" id (objStr target "gadget")) agent
    _ ->
      ledgerRoot ++ "/refs/" ++ safeToken (maybe "unknown" id (objStr target "target_ref")) ++ "/" ++ safeToken agent ++ ".json"

-- | Path of an agent's gadget replay ledger (used by the amalgamate audit too).
gadgetLedgerRelpath :: String -> String -> String -> String
gadgetLedgerRelpath gizmo gadget agent =
  ledgerRoot ++ "/gadgets/" ++ safeToken gizmo ++ "/" ++ safeToken gadget ++ "/" ++ safeToken (normalizeAgent agent) ++ ".json"

-- fingerprints ----------------------------------------------------------------

fileFingerprint :: Value -> Value
fileFingerprint item =
  let op = maybe "upsert" id (objStr item "op")
      path = maybe "" id (objStr item "path")
      base = ["op" .= op, "path" .= path]
  in if op == "upsert"
       then
         let content = maybe "" id (objStr item "content")
             enc = maybe "utf-8" id (objStr item "encoding")
         in object (base ++ [ "encoding" .= enc, "content_sha256" .= sha256Text content, "content_length" .= length content ])
       else object base

-- payload object --------------------------------------------------------------

writePayloadObject :: FilePath -> Value -> String -> IO String
writePayloadObject work payload sha = do
  let relpath = payloadRelpath sha
      path = work </> relpath
  exists <- doesFileExist path
  if exists
    then pure relpath
    else do
      now <- nowIso
      let obj = object ["format" .= payloadFormat, "json_patch_sha256" .= sha, "created_at" .= now, "payload" .= payload]
      createDirectoryIfMissing True (takeDirectory path)
      BL.writeFile path (encode obj)
      pure relpath

-- ledger read -----------------------------------------------------------------

emptyLedger :: String -> Value -> IO Value
emptyLedger agent target = do
  now <- nowIso
  pure $ object
    [ "format" .= ledgerFormat
    , "agent" .= agent
    , "target" .= target
    , "created_at" .= now
    , "updated_at" .= now
    , "next_sequence" .= (1 :: Int)
    , "entries" .= ([] :: [Value])
    ]

readLedger :: FilePath -> String -> Value -> IO Value
readLedger path agent target = do
  exists <- doesFileExist path
  if not exists
    then emptyLedger agent target
    else do
      decoded <- eitherDecodeFileStrict' path :: IO (Either String Value)
      case decoded of
        Right v@(Object _) -> pure v
        _ -> emptyLedger agent target

-- append ----------------------------------------------------------------------

-- | Append a landed JSON patch to its agent ledger (idempotent on json_patch_sha256).
-- Writes the payload object and the ledger into @work@. Returns the ledger relpath.
appendEntry :: FilePath -> String -> Value -> String -> IO String
appendEntry work agent0 payload targetRef = do
  let agent = normalizeAgent agent0
      target = targetDescriptor payload targetRef
      relpath = ledgerRelpath agent target
      path = work </> relpath
      files = asArray (objLookup payload "files")
      fps = map fileFingerprint files
      sha = jsonSha256 payload
      title = maybe "" id (objStr payload "title")
  payloadPath <- writePayloadObject work payload sha
  ledger <- readLedger path agent target
  let entries = asArray (objLookup ledger "entries")
      already = any (\e -> objStr e "json_patch_sha256" == Just sha) entries
  if already
    then pure relpath
    else do
      now <- nowIso
      let nextSeq = maybe (length entries + 1) id (objLookup ledger "next_sequence" >>= asInt)
          entry = object
            [ "format" .= entryFormat
            , "sequence" .= nextSeq
            , "agent" .= agent
            , "target" .= target
            , "title" .= title
            , "json_patch_sha256" .= sha
            , "payload_path" .= payloadPath
            , "file_count" .= length files
            , "file_fingerprints" .= fps
            , "target_ref_at_capture" .= targetRef
            , "created_at" .= now
            ]
          ledger' = object
            [ "format" .= ledgerFormat
            , "agent" .= agent
            , "target" .= target
            , "created_at" .= maybe now id (objStr ledger "created_at")
            , "updated_at" .= now
            , "next_sequence" .= (nextSeq + 1)
            , "entries" .= (entries ++ [entry])
            ]
      createDirectoryIfMissing True (takeDirectory path)
      BL.writeFile path (encode ledger')
      pure relpath
